---
params:
  sample_name: "2021_31"
title: "NF1 long-read"
subtitle: "Individual dataset"
author: "Ali Hamraoui & Audrey Onfroy"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    code_folding: show
    code_download: true
    toc: true
    toc_float: true
    number_sections: false
---

<style>
body {
text-align: justify}
</style>

<!-- Automatically computes and prints in the output the running time for any code chunk -->
```{r, echo=FALSE}
# https://github.com/rstudio/rmarkdown/issues/1453
hooks = knitr::knit_hooks$get()
hook_foldable = function(type) {
  force(type)
  function(x, options) {
    res = hooks[[type]](x, options)
    
    if (isFALSE(options[[paste0("fold_", type)]])) return(res)
    
    paste0(
      "<details><summary>", "show", "</summary>\n\n",
      res,
      "\n\n</details>"
    )
  }
}
knitr::knit_hooks$set(
  output = hook_foldable("output"),
  plot = hook_foldable("plot")
)
```

<!-- Set default parameters for all chunks -->
```{r, setup, include = FALSE}
set.seed(1)
knitr::opts_chunk$set(echo = TRUE, # display code
                      # display chunk output
                      message = FALSE,
                      warning = FALSE,
                      fold_output = FALSE, # for sessionInfo()
                      fold_plot = FALSE,
                      
                      # figure settings
                      fig.align = 'center',
                      fig.width = 20,
                      fig.height = 15)
```


The goal of this script is to generate a Seurat object for sample `r params$sample_name`.

* removal of cells based on quality control metrics
* normalization with `LogNormalize`, for only the remaining cells
* cell cycle and cell type annotation
* dimension reduction using `PCA`
* projection using `tSNE` and `UMAP`


```{r library}
library(dplyr)
library(patchwork)
library(ggplot2)

.libPaths()
```

# Preparation

In this section, we set the global settings of the analysis. We will store data there :

```{r out_dir}
out_dir = "."
```

We load custom functions :

```{r source_functions}
source(paste0(out_dir, "/../1_metadata/sourceable_functions.R"))
source(paste0(out_dir, "/../2_individual/load_sc_data.R"))

lapply(ls(), FUN = function(one_element) {
  if (class(eval(parse(text = one_element))) == "function") {
    out = one_element
  } else {
    out = NULL
  }
  return(out)
}) %>% unlist()
```

We load the parameters :

```{r get_param}
#sample_name = params$sample_name # "TOTO"
sample_name = "2020_26"
```

Input count matrices are there :

```{r count_matrix_dir}
count_matrix_dir = paste0(out_dir, "/input/", sample_name)
```

We load the markers and specific colors for each cell type :

```{r cell_markers}
cell_markers = readRDS(paste0(out_dir, "/../1_metadata/nf1_cell_markers.rds"))
cell_markers = lapply(cell_markers, FUN = toupper)
lengths(cell_markers)
```

Here are custom colors for each cell type :

```{r color_markers, fig.width = 12, fig.height = 1, class.source = "fold-hide"}
color_markers = readRDS(paste0(out_dir, "/../1_metadata/nf1_color_markers.rds"))

data.frame(cell_type = names(color_markers),
           color = unlist(color_markers)) %>%
  ggplot2::ggplot(., aes(x = cell_type, y = 0, fill = cell_type)) +
  ggplot2::geom_point(pch = 21, size = 5) +
  ggplot2::scale_fill_manual(values = unlist(color_markers), breaks = names(color_markers)) +
  ggplot2::theme_classic() +
  ggplot2::theme(legend.position = "none",
                 axis.line = element_blank(),
                 axis.title = element_blank(),
                 axis.ticks = element_blank(),
                 axis.text.y = element_blank())
```


We load markers to display on the dotplot :

```{r dotplot_markers}
dotplot_markers = readRDS(paste0(out_dir, "/../1_metadata/nf1_dotplot_markers.rds"))
dotplot_markers
```


We load metadata for this sample :

```{r sample_info}
sample_info = readRDS(paste0(out_dir, "/../1_metadata/nf1_sample_info.rds"))
sample_info %>%
  dplyr::filter(project_name == sample_name)
```


These is a parameter for different functions :

```{r global_settings}
cut_log_nCount_RNA = 6
cut_nFeature_RNA = 600
cut_percent.mt = 15
cut_percent.rb = 30
```

# Load count matrix

## Count matrices

In this section, we load the count matrices.

### Genes

We load the genes count matrix :

load("data/illumina_data.Rdata")
gene.mtx = Illumina$scNaUmi_seq

```{r load_genes_cm}
scmat_gene <- load_sc_data(data_path = paste0(count_matrix_dir, "/genes/raw_feature_bc_matrix"),
                           sample_name = sample_name,
                           my_seed = 1337L)
```

### Isoforms

```{r load_iso_cm}
scmat_iso <- Seurat::Read10X(paste0(count_matrix_dir, "/isoforms/isoform_matrix/"))
```

### Homogeneization

We homogenize both matrix annotations :

```{r homogenize}
colnames(scmat_gene) = gsub("-1$", "", colnames(scmat_gene))
com.cells = intersect(colnames(scmat_iso), colnames(scmat_gene))

scmat_iso = scmat_iso[, colnames(scmat_iso)%in%com.cells]
scmat_gene = scmat_gene[, colnames(scmat_gene)%in%com.cells]
```

## Make Seurat object

We create assays for each matrix :

```{r create_assays}
gene_assay = Seurat::CreateAssayObject(counts = scmat_gene)
rm(scmat_gene)
isoform_assay = Seurat::CreateAssayObject(counts = scmat_iso)
```

and initialize a Seurat object :

```{r init_sobj}
assay = "RNA"
seurat_obj = Seurat::CreateSeuratObject(counts = gene_assay,
                                        assay = assay,
                                        project.name = sample_name)

seurat_obj[["ISO"]] = isoform_assay
seurat_obj[[paste0('log_nCount_', assay)]] = log(seurat_obj[[paste0('nCount_', assay)]])

seurat_obj
head(rownames(seurat_obj@assays$RNA$counts))
head(rownames(seurat_obj@assays$ISO$counts))
```

We set an upper bound to 20,000 cells. Remaining cells correspond to ambiant RNA.

```{r filter_by_umi}
if (ncol(seurat_obj) > 20000) {
  seurat_obj$nb_UMI = Seurat::GetAssayData(seurat_obj, assay = "RNA", slot = "counts") %>%
    Matrix::colSums()
  
  thresh_20k = sort(seurat_obj$nb_UMI, decreasing = TRUE)[20000]
  
  seurat_obj = subset(seurat_obj, nb_UMI >= thresh_20k)
  sobj$nb_UMI = NULL
  
  seurat_obj
}
```

In genes metadata, we add the Ensembl ID. The `seurat_obj@assays$RNA@meta.features` dataframe contains three information :

* `rownames` : gene names stored as the dimnames of the count matrix. Duplicated gene names will have a `.1` at the end of their name
* `Ensembl_ID` : EnsemblID, as stored in the `features.tsv.gz` file
* `gene_name` : gene_name, as stored in the `features.tsv.gz` file. Duplicated gene names will have the same name.

```{r features_df}
features_df = read.csv(paste0(count_matrix_dir, "/genes/raw_feature_bc_matrix/features.tsv.gz"), sep = "\t", header = 0)
features_df = features_df[, c(1:2)]
colnames(features_df) = c("Ensembl_ID", "gene_name")
rownames(features_df) = rownames(seurat_obj) # mandatory for Seurat::FindVariableFeatures

seurat_obj@assays$RNA@meta.features = features_df
rm(features_df)

head(seurat_obj@assays$RNA@meta.features)
```

We add the same columns as in metadata :

```{r add_metadata}
row_oi = (sample_info$project_name == sample_name)

seurat_obj$project_name = sample_name
seurat_obj$sample_identifier = sample_info[row_oi, "sample_identifier"]
seurat_obj$sample_type = sample_info[row_oi, "sample_type"]

colnames(seurat_obj@meta.data)
```

# Before filtering

We generate a projection based on the genes count matrix, before cells filtering.

```{r set_assay_before}
Seurat::DefaultAssay(seurat_obj) = "RNA"
```

## Normalization

We normalize the genes count matrix.

```{r normalization}
seurat_obj = Seurat::NormalizeData(seurat_obj,
                                   normalization.method = "LogNormalize",
                                   assay = "RNA")

seurat_obj = Seurat::FindVariableFeatures(seurat_obj,
                                          assay = "RNA",
                                          normalization.method = "vst",
                                          nfeatures = 3000)
seurat_obj
```

## Projection

We generate a UMAP to visualize cells before filtering. First, we perform a PCA :

```{r pca_before, fig.width = 12, fig.height = 4}
seurat_obj = Seurat::ScaleData(seurat_obj,
                               features = rownames(seurat_obj), verbose = F)
var_features = Seurat::VariableFeatures(object = seurat_obj)

seurat_obj = Seurat::RunPCA(seurat_obj, features = var_features,
                    verbose = F, reduction.name = "RNA_pca", max_dims = 100)

Seurat::ElbowPlot(seurat_obj, ndims=100, reduction = "RNA_pca")
```

We generate a UMAP with 10 principal components :

```{r umap_before}
ndims = 20
seurat_obj = Seurat::RunUMAP(seurat_obj,
                             reduction = "RNA_pca",
                             dims = 1:ndims,
                             seed.use = 1337L,
                             reduction.name = paste0("RNA_pca_", ndims, "_umap"))

seurat_obj = Seurat::RunTSNE(seurat_obj,
                             reduction = "RNA_pca",
                             dims = 1:ndims,
                             seed.use = 1337L,
                             reduction.name = paste0("RNA_pca_", ndims, "_tsne"))

seurat_obj
```

## Cell type

We annotate cells for cell type using `Seurat::AddModuleScore` function.

```{r cell_annot_custom2}
seurat_obj = annotate.aud(seurat_obj, cell_markers)

colnames(seurat_obj@meta.data) = stringr::str_replace_all(string = colnames(seurat_obj@meta.data),
                                                          pattern = " ",
                                                          replacement = "_")

seurat_obj$cell_type = factor(seurat_obj$cell_type, levels = names(cell_markers))

table(seurat_obj$cell_type)
```

We visualize cell type annotation on the projection :

```{r umap_celltype, fig.height = 6, fig.width = 14}
fig <- list()
for (proj in c('umap', 'tsne')){
fig[[proj]] <- Seurat::DimPlot(seurat_obj, reduction = paste0("RNA_pca_20_", proj), group.by = "cell_type") +
  ggplot2::scale_color_manual(values = unlist(color_markers),
                              breaks = names(color_markers)) +
  ggplot2::labs(title = paste0(proj, "\n", sample_name),
                subtitle = paste0(ncol(seurat_obj), " cells")) +
  Seurat::NoLegend() + Seurat::NoAxes() +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_text(hjust = 0.5),
                 plot.subtitle = element_text(hjust = 0.5))
}
fig[["umap"]] | fig[["tsne"]]
rm(fig)
```

# Quality control

In this section, we look at the number of genes expressed by each cell, the number of UMI, the percentage of mitochondrial genes expressed, and the percentage of ribosomal genes expressed.

We compute four quality metrics :

```{r qc_metrics}
seurat_obj = Seurat::PercentageFeatureSet(seurat_obj, pattern = "^mt", col.name = "percent.mt")
seurat_obj = Seurat::PercentageFeatureSet(seurat_obj, pattern = "^Rp[l|s][0-9]*$", col.name = "percent.rb")
head(seurat_obj@meta.data)
```

We get the cell barcodes for the failing cells :

```{r failed}
fail_percent.mt = seurat_obj@meta.data %>% dplyr::filter(percent.mt > cut_percent.mt) %>% rownames()
fail_percent.rb = seurat_obj@meta.data %>% dplyr::filter(percent.rb > cut_percent.rb) %>% rownames()
fail_log_nCount_RNA = seurat_obj@meta.data %>% dplyr::filter(log_nCount_RNA < cut_log_nCount_RNA) %>% rownames()
fail_nFeature_RNA = seurat_obj@meta.data %>% dplyr::filter(nFeature_RNA < cut_nFeature_RNA) %>% rownames()
```

## Quality control representation

We can visualize the 4 cells quality with a Venn diagram : 

```{r qc_venn, class.source = "fold-hide", fig.width = 8, fig.height = 6}
n_filtered = c(fail_percent.mt, fail_percent.rb, fail_log_nCount_RNA, fail_nFeature_RNA) %>%
  unique() %>% length()
percent_filtered = round(100*(n_filtered/ncol(seurat_obj)), 2)

ggvenn::ggvenn(list(percent.mt = fail_percent.mt,
                    percent.rb = fail_percent.rb,
                    log_nCount_RNA = fail_log_nCount_RNA,
                    nFeature_RNA = fail_nFeature_RNA), 
               fill_color = c("#0073C2FF", "#EFC000FF", "orange", "pink"),
               stroke_size = 0.5, set_name_size = 4) +
  ggplot2::labs(title = "Filtered out cells",
                subtitle = paste0(n_filtered, " cells (", percent_filtered, " % of all cells)")) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, face = "bold"),
                 plot.subtitle = element_text(hjust = 0.5))
```


### Number of UMI

```{r vln_umi_cell_type, fig.width = 12, fig.height = 6}
Seurat::VlnPlot(seurat_obj, features = "log_nCount_RNA", pt.size = 0.001,
                group.by = "cell_type", cols = color_markers) +
  ggplot2::scale_fill_manual(values = color_markers, breaks = names(color_markers)) +
  ggplot2::geom_hline(yintercept = cut_log_nCount_RNA, col = "red") +
  ggplot2::labs(x = "")
```

```{r qc_umi_proj, fig.width = 8, fig.height = 7}
seurat_obj$fail = ifelse(colnames(seurat_obj) %in% fail_log_nCount_RNA,
                         yes = as.character(seurat_obj$cell_type), no = NA)
seurat_obj$fail = factor(seurat_obj$fail, levels = c(levels(seurat_obj$cell_type), NA))

Seurat::DimPlot(seurat_obj, group.by = "fail", na.value = "gray80", cols = color_markers) +
  ggplot2::labs(title = "log_nCount_RNA",
                subtitle = paste0(length(fail_log_nCount_RNA), " cells")) +
  Seurat::NoAxes() +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_text(hjust = 0.5),
                 plot.subtitle = element_text(hjust = 0.5))
```

### Number of features

```{r vln_features_cell_type, fig.width = 12, fig.height = 6}
Seurat::VlnPlot(seurat_obj, features = "nFeature_RNA", pt.size = 0.001,
                group.by = "cell_type", cols = color_markers) +
  ggplot2::scale_fill_manual(values = color_markers, breaks = names(color_markers)) +
  ggplot2::geom_hline(yintercept = cut_nFeature_RNA, col = "red") +
  ggplot2::labs(x = "")
```

```{r qc_features_proj, fig.width = 8, fig.height = 7}
seurat_obj$fail = ifelse(colnames(seurat_obj) %in% fail_nFeature_RNA,
                         yes = as.character(seurat_obj$cell_type), no = NA)
seurat_obj$fail = factor(seurat_obj$fail, levels = c(levels(seurat_obj$cell_type), NA))

Seurat::DimPlot(seurat_obj, group.by = "fail", na.value = "gray80", cols = color_markers) +
  ggplot2::labs(title = "nFeature_RNA",
                subtitle = paste0(length(fail_nFeature_RNA), " cells")) +
  Seurat::NoAxes() +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_text(hjust = 0.5),
                 plot.subtitle = element_text(hjust = 0.5))
```

### Mitochondrial genes expression

```{r vln_percentmt_cell_type, fig.width = 12, fig.height = 6}
Seurat::VlnPlot(seurat_obj, features = "percent.mt", pt.size = 0.001,
                group.by = "cell_type", cols = color_markers) +
  ggplot2::scale_fill_manual(values = color_markers, breaks = names(color_markers)) +
  ggplot2::geom_hline(yintercept = cut_percent.mt, col = "red") +
  ggplot2::labs(x = "")
```

```{r qc_mito_proj, fig.width = 8, fig.height = 7}
seurat_obj$fail = ifelse(colnames(seurat_obj) %in% fail_percent.mt,
                         yes = as.character(seurat_obj$cell_type), no = NA)
seurat_obj$fail = factor(seurat_obj$fail, levels = c(levels(seurat_obj$cell_type), NA))

Seurat::DimPlot(seurat_obj, group.by = "fail", na.value = "gray80", cols = color_markers) +
  ggplot2::labs(title = "percent.mt",
                subtitle = paste0(length(fail_percent.mt), " cells")) +
  Seurat::NoAxes() +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_text(hjust = 0.5),
                 plot.subtitle = element_text(hjust = 0.5))
```

### Ribosomal genes expression

```{r vln_percentrb_cell_type, fig.width = 12, fig.height = 6}
Seurat::VlnPlot(seurat_obj, features = "percent.rb", pt.size = 0.001,
                group.by = "cell_type", cols = color_markers) +
  ggplot2::scale_fill_manual(values = color_markers, breaks = names(color_markers)) +
  ggplot2::geom_hline(yintercept = cut_percent.rb, col = "red") +
  ggplot2::labs(x = "")
```

```{r qc_ribo_proj, fig.width = 8, fig.height = 7}
seurat_obj$fail = ifelse(colnames(seurat_obj) %in% fail_percent.rb,
                         yes = as.character(seurat_obj$cell_type), no = NA)
seurat_obj$fail = factor(seurat_obj$fail, levels = c(levels(seurat_obj$cell_type), NA))

Seurat::DimPlot(seurat_obj, group.by = "fail", na.value = "gray80", cols = color_markers) +
  ggplot2::labs(title = "percent.rb",
                subtitle = paste0(length(fail_percent.rb), " cells")) +
  Seurat::NoAxes() +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_text(hjust = 0.5),
                 plot.subtitle = element_text(hjust = 0.5))
```

# Filtering

We remove :

* cells with a number of UMI lower than `r cut_log_nCount_RNA`
* cells expressing a number of genes lower than `r cut_nFeature_RNA`
* cells having more than `r cut_percent.mt` \% of UMI related to mitochondrial genes
* cells having more than `r cut_percent.rb` \% of UMI related to ribosomal genes


```{r filter_cells}
seurat_obj = subset(seurat_obj, invert = TRUE,
                    cells = unique(c(fail_log_nCount_RNA, fail_nFeature_RNA,
                                     fail_percent.mt, fail_percent.rb)))
seurat_obj
```

```{r clean_filter, echo = FALSE}
rm(fail_percent.mt, fail_percent.rb, fail_log_nCount_RNA, fail_nFeature_RNA,
   cut_percent.mt, cut_percent.rb, cut_log_nCount_RNA, cut_nFeature_RNA)
```


# Post-filtering processing

We process the filtered dataset, starting with the genes count matrix.

## Normalization

We normalize the genes count matrix.

```{r normalization2}
seurat_obj = Seurat::NormalizeData(seurat_obj,
                                   normalization.method = "LogNormalize",
                                   assay = "RNA")

seurat_obj = Seurat::FindVariableFeatures(seurat_obj,
                                          assay = "RNA",
                                          normalization.method = "vst",
                                          nfeatures = 3000)
seurat_obj
```

## Projection

We generate a UMAP to visualize cells before filtering. First, we perform a PCA :

```{r pca_after, fig.width = 12, fig.height = 4}
seurat_obj = Seurat::ScaleData(seurat_obj,
                               features = rownames(seurat_obj), verbose = F)
var_features = Seurat::VariableFeatures(object = seurat_obj)

seurat_obj = Seurat::RunPCA(seurat_obj, features = var_features,
                            assay = "RNA",
                            verbose = F, 
                            reduction.name = "RNA_pca", max_ndims=100)

Seurat::ElbowPlot(seurat_obj, ndims= 100, reduction = "RNA_pca")
```

We generate a UMAP with 10 principal components :

```{r umap_after}
ndims = 20
seurat_obj = Seurat::RunUMAP(seurat_obj,
                             reduction = "RNA_pca",
                             dims = 1:ndims,
                             seed.use = 1337L,
                             reduction.name = paste0("RNA_pca_", ndims, "_umap"))

seurat_obj = Seurat::RunTSNE(seurat_obj,
                             reduction = "RNA_pca",
                             dims = 1:ndims,
                             seed.use = 1337L,
                             reduction.name = paste0("RNA_pca_", ndims, "_tsne"))

seurat_obj
```

## Cell type

We annotate cells for cell type using `Seurat::AddModuleScore` function.

```{r cell_annot_custom}
seurat_obj = annotate.aud(seurat_obj, cell_markers)

colnames(seurat_obj@meta.data) = stringr::str_replace_all(string = colnames(seurat_obj@meta.data),
                                                          pattern = " ",
                                                          replacement = "_")

seurat_obj$cell_type = factor(seurat_obj$cell_type, levels = names(cell_markers))

table(seurat_obj$cell_type)
```

We visualize cell type annotation on the projection :

```{r umap_celltype_filtered, fig.height = 6, fig.width = 12}
cell_type <- list()
for (proj in c("tsne", "umap")){
cell_type[[proj]] = Seurat::DimPlot(seurat_obj, reduction = paste0("RNA_pca_20_", proj), group.by = "cell_type") +
  ggplot2::scale_color_manual(values = unlist(color_markers),
                              breaks = names(color_markers)) +
  ggplot2::labs(title = paste0(proj,"\n",sample_name),
                subtitle = paste0(ncol(seurat_obj), " cells")) +
  Seurat::NoLegend() + Seurat::NoAxes() +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_text(hjust = 0.5),
                 plot.subtitle = element_text(hjust = 0.5))
}
cell_type[["umap"]] | cell_type[["tsne"]]
```

## Clustering

We generate a clustering :

```{r clustering}
seurat_obj = Seurat::FindNeighbors(seurat_obj, reduction = "RNA_pca", dims = c(1:ndims))
seurat_obj = Seurat::FindClusters(seurat_obj, resolution = 0.5)

table(seurat_obj$seurat_clusters)
```

We visualize the clustering :

```{r see_clusters, fig.width = 12, fig.height = 6}
clusters = Seurat::DimPlot(seurat_obj, group.by = "seurat_clusters", label = TRUE,
                           reduction = paste0("RNA_pca_", ndims, "_umap")) +
  Seurat::NoAxes() + ggplot2::ggtitle("Clustering") +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_text(hjust = 0.5))

clusters | cell_type[["umap"]]
```

## Isoforms processing

Here, we are looking for events where two isoforms of the same gene are considered markers of different clusters.

We process the isoforms count matrix :

```{r process_iso}
seurat_obj = Seurat::NormalizeData(seurat_obj, assay = "ISO",
                                   normalization.method = "LogNormalize", verbose = F) 
seurat_obj = Seurat::ScaleData(seurat_obj, assay = "ISO", verbose = FALSE)
```

We generate a "multi" assay using `isoswitch` package :

```{r isoswitch_multi}
seurat_obj = isoswitch::iso_preprocess(seurat_obj, assay = "ISO", new_assay = "multi", filter_threshold = 5)
head(rownames(seurat_obj@assays$multi@counts))
```

We finally have three assays with distinct dimensions :

```{r dim_assays}
dim(seurat_obj@assays$RNA@counts)
dim(seurat_obj@assays$ISO@counts)
dim(seurat_obj@assays$multi@counts)
```

we search for gene markers. then returns genes with >1 transcripts in different clusters. (ISO_SWITCH_ALL() method uses Seurat’s FindMarkers)

```{r iso_compute_stats}
stats = isoswitch::iso_compute_stats(seurat_obj@assays$multi@counts) %>%
  dplyr::arrange(gene_id)

head(stats, n=4)
```

We plot a summary with number of genes, number of transcripts, distribution of isoforms and number of genes per cluster.

```{r plot_assay_stats, fig.height=6, fig.width=12}
isoswitch::plot_assay_stats(seurat_obj, "ISO")
```

# Visualisation

We visualize all cell types markers on the UMAP :

```{r plot_genes, fig.width = 12, fig.height = 20}
markers = dotplot_markers %>% unlist() %>% unname()
markers = markers[markers %in% rownames(seurat_obj[["RNA"]])]

plot_list = lapply(markers,
                   FUN = function(one_gene) {
                     p = Seurat::FeaturePlot(seurat_obj, features = one_gene,
                                             reduction = paste0("RNA_pca_", ndims, "_umap")) +
                       ggplot2::labs(title = one_gene) +
                       ggplot2::scale_color_gradientn(colors = c("lightgray", "#FDBB84", "#EF6548", "#7F0000", "black")) +
                       ggplot2::theme(aspect.ratio = 1,
                                      plot.subtitle = element_text(hjust = 0.5)) +
                       Seurat::NoAxes()
                     return(p)
                   })

patchwork::wrap_plots(plot_list, ncol = 4)
```
# Save

We save the annotated and filtered Seurat object :

```{r save_seurat_obj_filtered_annotated}
saveRDS(seurat_obj, file = paste0(out_dir, "/datasets/", sample_name, "_seurat_obj_filtered.rds"))
```

# R session

```{r sessioninfo, echo = FALSE, fold_output = TRUE}
sessionInfo()
```
